nifi-toolkit/nifi-toolkit-cli/src/main/java/org/apache/nifi/toolkit/cli/impl/command/registry/flow/ExportAllFlows.java

```diff
@@ -0,0 +1,182 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class ExportAllFlows extends AbstractNiFiRegistryCommand<VersionedFlowSnapshotsResult> {
+    private static final String ALL_BUCKETS_COLLECTED = "All buckets collected...";
+    private static final String ALL_FLOWS_COLLECTED = "All flows collected...";
+    private static final String ALL_FLOW_VERSIONS_COLLECTED = "All flow versions collected...";
+    private static final String EXPORTING_FLOW_VERSIONS = "Exporting flow versions...";
+    private final ListBuckets listBuckets;
+    private final ListFlows listFlows;
+    private final ListFlowVersions listFlowVersions;
+    private final ExportFlowVersion exportFlowVersion;
+
+    public ExportAllFlows() {
+        super("export-all-flows", VersionedFlowSnapshotsResult.class);
+        this.listBuckets = new ListBuckets();
+        this.listFlows = new ListFlows();
+        this.listFlowVersions = new ListFlowVersions();
+        this.exportFlowVersion = new ExportFlowVersion();
+    }
+
+    @Override
+    public void doInitialize(final Context context) {
+        addOption(CommandOption.OUTPUT_DIR.createOption());
+
+        listBuckets.initialize(context);
+        listFlows.initialize(context);
+        listFlowVersions.initialize(context);
+        exportFlowVersion.initialize(context);
+    }
+
+    @Override
+    public String getDescription() {
+        return "List all the buckets, for each bucket, list all the flows, for each flow, list all versions and export each version." +
+                "Versions will be saved in the provided target directory.";
+    }
+
+    @Override
+    public VersionedFlowSnapshotsResult doExecute(NiFiRegistryClient client, Properties properties) throws IOException, NiFiRegistryException, ParseException, CommandException {
+        final String outputDirectory = getRequiredArg(properties, CommandOption.OUTPUT_DIR);
+        final boolean isInteractive = getContext().isInteractive();
+
+        // Gather all buckets and create a map for quick access by bucket id
+        final Map<String, Bucket> bucketMap = getBucketMap(client, isInteractive);
+
+        // Gather all flows and create a map for quick access by flow id
+        final Map<String, VersionedFlow> flowMap = getFlowMap(client, bucketMap, isInteractive);
+
+        // Gather all versions for all the flows
+        final List<VersionedFlowSnapshotMetadata> versionedFlowSnapshotMetadataList = getVersionedFlowSnapshotMetadataList(client, flowMap, isInteractive);
+
+        // Prepare flow version exports
+        final List<VersionedFlowSnapshot> versionedFlowSnapshotList = getVersionedFlowSnapshotResults(client, outputDirectory, bucketMap, flowMap, versionedFlowSnapshotMetadataList, isInteractive);
```

Comment:
Are there potential memory concerns with this approach? What happens if there are hundreds of large flows?

Reply:
I think the way the result classes worked kind of led it to be implemented this way, but I agree with David's concern that we should see if we can write out the snapshots one-by-one. Maybe we can pass some kind iterator/helper object to the result class that allow it to keep calling next() and writing out the result?